### PR TITLE
Modernization - PHP 7.1, strict types, Build Stages, dropped legacy adapters etc.

### PR DESCRIPTION
**Pull request by Majkl578**
This PR includes bunch of changes, some of them BC break-ish:
* PHP 7.1 is now required, HHVM is dropped - matches other Doctrine packages
* tested against PHP 7.2
* removed Coveralls, obsoleted in other packages - should probably be replaced by Scrutinizer, but I don't have permissions to set that up (anyone?)
* migrated to ext/mongodb - ext/mongo is PHP 5 only
* dropped Riak cache support - not compatible with PHP 7 - not a real BC break
* dropped Memcache cache support - not compatible with PHP 7 - not a real BC break
* upgraded PHPUnit
* replaced external Coding Standard with Doctrine's, but not applied/enforced as it'd be a lot of changes - could be done if no objections about conflicts are raised
* enabled strict types - possible BC break, nothing revealed by tests though
* removed licence headers - matches other Doctrine packages

Future scope:
* PHPStan
* apply and enforce Doctrine CS

Test failure under 7.2 looks like a redis extension bug. :/

### CODE REVIEW

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -186,13 +171,11 @@ public function getConfigTreeBuilder()
                             ->append($this->addCouchbaseNode())
                             ->append($this->addChainNode())
                             ->append($this->addMemcachedNode())
-                            ->append($this->addMemcacheNode())
```
Comment (alcaeus) [inline DependencyInjection/Configuration.php]: Wouldn't this invalidate existing config with a `memcache` key?

File: DependencyInjection/Definition/MemcacheDefinition.php [hunk 2]
```diff
@@ -1,62 +0,0 @@
...
-use Symfony\Component\DependencyInjection\ContainerBuilder;
-use Symfony\Component\DependencyInjection\Definition;
-use Symfony\Component\DependencyInjection\Reference;
-
-/**
- * Memcache definition.
- *
- * @author Fabio B. Silva <fabio.bat.silva@gmail.com>
- */
-class MemcacheDefinition extends CacheDefinition
```
Comment (alcaeus) [inline DependencyInjection/Definition/MemcacheDefinition.php]: Dropping a class is most definitely a BC break.

**Comment by lcobucci** [comment]
Although I REALLY like the idea of modernising all our packages I think we shouldn't do it on the bundles/modules (I actually believe that we should transfer them to SF/ZF organisations, but that's a different story).

What I mean is that we did that kind of work on another bundle and SF team is reverting some changes to ensure compatibility with SF and I want to avoid that kind of thing.

**Comment by alcaeus** [comment]
@lcobucci good point - let’s ask @fabpot how he wants to proceed.

**Comment by alcaeus** [comment]
Ping @fabpot @weaverryan is it ok to drop PHP 5 for cache-bundle or would you rather keep it (for now)?

**Comment by fabpot** [comment]
I'm fine with dropping PHP 5 as long as there is still a maintained version for bug fixed on 1.3 which would keep PHP 5 compat.

**Comment by alcaeus** [comment]
Closing here - this will be introduced in multiple Mrs, starting with dropping support for PHP < 7.1 in #147.

**Pull request closed without merging**